cmd/build_image.go

```diff
@@ -0,0 +1,268 @@
+			Required:    false,
+			Destination: &c.RepoURL,
+		},
+		&cli.StringFlag{
+			Name: "repo-ref",
+			Usage: "Git reference to use (mutually exclusive with version), if set, repo will be cloned to a sub directory of sdk-features." +
+				"Used as an image tag.",
+			Required:    false,
+			Destination: &c.RepoRef,
+		},
+		&cli.StringFlag{
+			Name:        "platform",
+			Usage:       "Platform to build the container for (docker build --platform)",
+			Required:    false,
+			Destination: &c.Platform,
+		},
+		&cli.StringFlag{
+			Name:        "image-name",
+			Usage:       "Name of the image to build",
+			Required:    false,
+			DefaultText: DEFAULT_IMAGE_NAME,
+			Destination: &c.ImageName,
+		},
+	}
+}
+
+// ImageBuilder builds docker images.
+type ImageBuilder struct {
+	log     log.Logger
+	config  ImageBuildConfig
+	rootDir string
+}
+
+// NewImageBuilder creates a new builder for the given config.
+func NewImageBuilder(config ImageBuildConfig) *ImageBuilder {
+	return &ImageBuilder{
+		// TODO(cretz): Configurable logger
+		log:     log.NewCLILogger(),
+		config:  config,
+		rootDir: rootDir(),
+	}
+}
+
+// BuildImage builds a docker image based on builder config.
+func (i *ImageBuilder) BuildImage(ctx context.Context) error {
+	var err error
+	if i.config.Lang, err = normalizeLangName(i.config.Lang); err != nil {
+		return err
+	}
+
+	// Build from one of: repo or version
+	if i.config.RepoRef != "" {
+		if i.config.Version != "" {
+			return fmt.Errorf("version and repo-ref are mutually exclusive")
+		}
+		return i.buildFromRepo(ctx)
+	} else if i.config.Version != "" {
+		return i.buildFromVersion(ctx)
+	} else {
+		return fmt.Errorf("either version or repo-ref is required")
+	}
+}
+
+func (i *ImageBuilder) buildFromRepo(ctx context.Context) error {
+	if i.config.RepoURL == "" {
+		var err error
+		i.config.RepoURL, err = defaultRepoURL(i.config.Lang)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We have to clone into rootDir because it's part of the docker context
+	tempDir, err := os.MkdirTemp(i.rootDir, "cloned-repo-*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	repoBaseDir, err := i.gitClone(ctx, tempDir)
+	if err != nil {
+		return err
+	}
+
+	repoDir := filepath.Join(filepath.Base(tempDir), repoBaseDir)
+
+	// Get the actual ref (in case passed in ref is a branch name or tag)
+	repoRef, err := i.gitRef(ctx, filepath.Join(tempDir, repoBaseDir, ".git"))
+	if err != nil {
+		return err
+	}
+
+	tags := []string{fmt.Sprintf("%s-%s", i.config.Lang, i.config.RepoRef)}
+
+	return i.dockerBuild(ctx, buildConfig{
+		tags:      tags,
+		labels:    map[string]string{"SDK_REPO_URL": i.config.RepoURL, "SDK_REPO_REF": repoRef},
+		buildArgs: map[string]string{"SDK_VERSION": repoDir, "REPO_DIR_OR_PLACEHOLDER": repoDir},
+	})
+}
+
+func (i *ImageBuilder) buildFromVersion(ctx context.Context) error {
+	version := semver.Canonical(i.config.Version)
```

Comment:
I don't think you should validate this and I don't think you should use the Go library which requires a `v` prefix. Can't we just use the version as the tag directly? We intentionally don't validate version in other SDK features commands.

Reply:
I need this for generating the different semver tags (1, 1.16, 1.16.0), All of our SDKs support semver except for Python for now (granted the `v` prefix is go specific).
I will probably end up changing this a bit when I implement building containers for another language.

Reply:
Ah, Docker tags? :+1:

Reply:
Yup